hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/AbstractJoiner.java

```diff
@@ -456,19 +456,23 @@ private boolean prepareClusterState(ClusterServiceImpl clusterService) {
         }
 
         long until = Clock.currentTimeMillis() + mergeNextRunDelayMs;
-        while (clusterService.getClusterState() == ClusterState.ACTIVE) {
-            try {
-                clusterService.changeClusterState(ClusterState.FROZEN);
+        while (Clock.currentTimeMillis() < until) {
+            ClusterState clusterState = clusterService.getClusterState();
+            if (!clusterState.isMigrationAllowed() && !clusterState.isJoinAllowed()
```

Comment:
Yes you're right. Some other process can change cluster state concurrently. Merge process is not atomic. I would not even try to make it :) Existing merge mechanism is mostly kept as is.